README.md
# IPTV Simple PVR
IPTV Live TV and Radio PVR client addon for [Kodi] (http://kodi.tv)

##### Useful links

* [Kodi's PVR user support] (http://forum.kodi.tv/forumdisplay.php?fid=167)
* [Kodi's PVR development support] (http://forum.kodi.tv/forumdisplay.php?fid=136)

# Modififed to build againist Krypton (RC3)


pvr.iptvsimple/changelog.txt
v1.12.5
- Updated to PVR API 5.2.1 (Krypton)

v1.12.4
- Updated to PVR API v3.0.0 (API 1.9.7 compatibility mode)

v1.12.3
- Updated to PVR API v2.1.0
- Automatically fill in platform and library name

v1.12.2
- Debian packaging: fix library install path
- Miscellaneous backend fixes

v1.12.1
- updated to API v2.0.0

v1.12.0
- updated to API v1.9.7

v1.11.3
- updated language files from Transifex

v1.11.2
- updated to use new libplatform-dev

v1.11.1
- fixed EPG loading.

v1.11.0
- updated to API v1.9.6
- new feature: channel logos from xmltv
- new feature: map xmltv genre to Kodi genre via genres.xml
- now possible to define full logo path in m3u if logos path is empty in setting
- a lot of code improvements.

v1.10.2
- updated to API v1.9.5

v1.10.1
- language files from Transifex
- minor changes to conform with C++11
- Updated to API v1.9.4

v1.9.12
- added getBackendHostname function

v1.9.10
- change library name to Kodi

v1.9.9
- fixed deadlock when activating the addon on some darwin (osx, ios) runtimes

v1.9.6
- updated to API v1.9.2

v1.9.3
- added setting to disable cache m3u/epg
- added support <icon src=""/> for programme in epg
- added possibility to specify logo path as URL
- changed tvg-logo should contains full logo filename (with extention)
- changed don't try download xmltv if it doesn't load after 3 attempt.

v1.9.2
- add timeshift buffer functions

v1.8.1
- fixed issue with BOM header in EPG XML
- fixed issue with update channels if they are changed in m3u
- added setting starting number of channels

v1.8.0
- Initial version
- Supports m3u as Live TV streams source and XMLTV as EPG source
- Supports Live TV, Radio channels and EPG


src/PVRRecorderThread.cpp
/*
 *      Copyright (C) 2015 Radek Kubera
 *      http://github.com/afedchin/xbmc-addon-iptvsimple/
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with XBMC; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */

#include "p8-platform/util/StringUtils.h"
#include <map>
#include "p8-platform/threads/threads.h"
#include "PVRRecorderThread.h"
#include "PVRIptvData.h"
#include "PVRRecJob.h"
#include "PVRSchedulerThread.h"
#include "PVRUtils.h"
#include "PVRPlayList.h"
#include "libexecstream/exec-stream.h"

#include <iostream>
#include <fstream>
#include <string>
#include <unistd.h>

using namespace ADDON;

extern PVRRecJob *p_RecJob;
extern PVRSchedulerThread *p_Scheduler;
extern string g_recordingsPath;
extern string g_ffmpegPath;
extern string g_ffmpegParams;
extern string g_rtmpdumpPath;
extern string g_fileExtension;
extern int g_streamTimeout;

extern bool s_triggerTimerUpdate;

PVRRecorderThread::PVRRecorderThread(PVRIptvChannel &currentChannel, int iClientIndex)
{
    isWorking = false;
    t_iClientIndex = iClientIndex;
    t_currentChannel = currentChannel;
    CreateThread();
}

PVRRecorderThread::~PVRRecorderThread(void)
{
    PVR_REC_JOB_ENTRY entry;
    PVRIptvChannel currentChannel;
    p_RecJob->getJobEntry(t_iClientIndex, entry);
        
    XBMC->Log(LOG_DEBUG,"Closing thread %s",entry.Timer.strTitle);
    while (entry.Status==PVR_STREAM_IS_RECORDING)
    {
        p_RecJob->getJobEntry(t_iClientIndex, entry);
        XBMC->Log(LOG_DEBUG,"Closing recording thread %s",entry.Timer.strTitle);
    }
}

void PVRRecorderThread::StopThread(bool bWait /*= true*/)
{
    PVR_REC_JOB_ENTRY entry;
    p_RecJob->getJobEntry(t_iClientIndex, entry);
    entry.Status = PVR_STREAM_IS_STOPPING;
    p_RecJob->updateJobEntry(entry);
    XBMC->Log(LOG_DEBUG,"Stopping thread %s",entry.Timer.strTitle);
    CThread::StopThread(bWait);
}

void PVRRecorderThread::CorrectDurationFLVFile (const string &videoFile, const double &duration)
{
    if (duration<0){
	XBMC->Log(LOG_NOTICE, "Duration correction failed");
	return;
    }
    //read 1024 first file bytes
    char buffer[1024];
    void *fileHandle;
    fileHandle = XBMC->OpenFile(videoFile.c_str(), 0);
    XBMC->ReadFile(fileHandle, buffer, 1024);
    XBMC->CloseFile(fileHandle);
    int loop_end = 1024-8-sizeof(double);
    int loop;
    int pos = -1;
    for (loop=0; loop<loop_end;loop++)
    {
        if (buffer[loop]=='d' && buffer[loop+1]=='u' && buffer[loop+2]=='r' && buffer[loop+3]=='a' && buffer[loop+4]=='t' && buffer[loop+5]=='i' && buffer[loop+6]=='o' && buffer[loop+7]=='n')
            pos = loop;
    }
    //correct 4096 first file bytes
    if (pos>=0)
    {
        pos = pos+9;
	union
	{
	    unsigned char dc[8];
	    double dd;
	} d;
	d.dd = duration;
	long one = 1;
	//is isBigEndian?
	if(!(*((char *)(&one))))
	{
	    buffer[pos+0] = d.dc[0]; buffer[pos+1] = d.dc[1]; buffer[pos+2] = d.dc[2]; buffer[pos+3] = d.dc[3]; buffer[pos+4] = d.dc[4]; buffer[pos+5] = d.dc[5]; buffer[pos+6] = d.dc[6]; buffer[pos+7] = d.dc[7];
	}
	else
	{
	    buffer[pos+0] = d.dc[7]; buffer[pos+1] = d.dc[6]; buffer[pos+2] = d.dc[5]; buffer[pos+3] = d.dc[4]; buffer[pos+4] = d.dc[3]; buffer[pos+5] = d.dc[2]; buffer[pos+6] = d.dc[1]; buffer[pos+7] = d.dc[0];
	}
        
	fileHandle = XBMC->OpenFileForWrite(videoFile.c_str(), 0);
	XBMC->SeekFile(fileHandle,0, ios::beg);
	int size = XBMC->WriteFile(fileHandle, buffer, 1024);
	if (size>0) {
	    XBMC->Log(LOG_NOTICE, "Duration corrected");
	}
	XBMC->CloseFile(fileHandle);
	return;
    }
    XBMC->Log(LOG_NOTICE, "Duration correction failed");
}

void *PVRRecorderThread::Process(void)
{   
    PVR_REC_JOB_ENTRY entry;
    p_RecJob->getJobEntry(t_iClientIndex, entry);
    entry.Status = PVR_STREAM_IS_RECORDING;
    p_RecJob->updateJobEntry(entry);
    
    struct tm *current;
    time_t now;
	
    time(&now);
    current = localtime(&now);
    string month = inttostr(current->tm_mon);
    if (current->tm_mon<10) month = "0"+month;
    string day = inttostr(current->tm_mday);
    if (current->tm_mday<10) day = "0"+day;
    string hour = inttostr(current->tm_hour);
    if (current->tm_hour<10) hour = "0"+hour;
    string min = inttostr(current->tm_min);
    if (current->tm_min<10) min = "0"+min;
    string sec = inttostr(current->tm_sec);
    if (current->tm_sec<10) sec = "0"+sec;
    
    string strDate = " ("+inttostr(current->tm_year+1900)+"-"+month+"-"+day+" "+hour+"-"+min+"-"+sec+")";
    
    XBMC->Log(LOG_NOTICE, "Try to open stream %s", t_currentChannel.strStreamURL.c_str());
    
    //GetPlayList
    vector<string> vstrList;
    PVRPlayList* playList = new PVRPlayList();
    string strStreamUrl = t_currentChannel.strStreamURL;
    playList->GetPlaylist (strStreamUrl, vstrList);
    delete (playList);
    
    string filename = entry.Timer.strTitle;
    filename = filename+strDate+"."+g_fileExtension;
    
    string illegalChars = "\\/:?\"<>|*'";
    string::iterator it ( filename.begin() );
    for (it = filename.begin() ; it < filename.end() ; ++it){
	bool found = illegalChars.find(*it) != string::npos;
	if(found){
	    *it = ' ';
	}
    }
    
    string videoFile = g_recordingsPath + filename;
    
    XBMC->Log(LOG_NOTICE,"File to write: %s ",videoFile.c_str());
    string strParams;
    string strCommand;
    
    double duration = entry.Timer.endTime-entry.Timer.startTime;
    t_duration = duration;
    
    double length = 0;
    
    vector<string> stremaUrlVect = StringUtils::Split (strStreamUrl," ");
    if (stremaUrlVect.size()>0) {
	strParams = stremaUrlVect[0];
	for (unsigned int i=1; i<stremaUrlVect.size();i++) {
	    string line = StringUtils::Trim(stremaUrlVect[i]);
	    if (line.length()>0) {
		vector<string> lineVect = StringUtils::Split (line,"=");
		if (lineVect[0]=="live") {
		    strParams = strParams + " --live";
		}
		else {
		    strParams = strParams + " --"+line;
		}
	    }
        }
    }
    
    if(strStreamUrl.substr(0, 7) == "rtmp://"
       || strStreamUrl.substr(0, 8) == "rtmpt://"
       || strStreamUrl.substr(0, 8) == "rtmpe://"
       || strStreamUrl.substr(0, 9) == "rtmpte://"
       || strStreamUrl.substr(0, 8) == "rtmps://")
    {
	strParams = " -r "+strParams;
	if (g_rtmpdumpPath.length()==0)
	{
	    XBMC->Log(LOG_ERROR,"Path to rtmpdump binary is not set. Please change addon configuration.");
	    entry.Status = PVR_STREAM_STOPPED;
            entry.Timer.state= PVR_TIMER_STATE_ERROR;
	    p_RecJob->updateJobEntry(entry);
	    s_triggerTimerUpdate = true;
	    return NULL;
	}
	   
	string strCommandLog = g_rtmpdumpPath+strParams;
	strCommand = g_rtmpdumpPath;
	XBMC->Log(LOG_NOTICE,"Starting rtmpdump: %s", strCommandLog.c_str());
    }
    else
    {
	strParams =  " -i \""+strParams+"\" "+g_ffmpegParams+" -f "+g_fileExtension+" -";
	if (g_ffmpegPath.length()==0)
	{
	    XBMC->Log(LOG_ERROR,"Path to ffmpeg binary is not set. Please change addon configuration.");
	    entry.Status = PVR_STREAM_STOPPED;
            entry.Timer.state= PVR_TIMER_STATE_ERROR;
	    p_RecJob->updateJobEntry(entry);
	    s_triggerTimerUpdate = true;
	    return NULL;
	}
	
	if (g_ffmpegParams.length()==0)
	{
	    XBMC->Log(LOG_ERROR,"Recompression params for ffmpeg are not set. Please change addon configuration.");
	    entry.Status = PVR_STREAM_STOPPED;
            entry.Timer.state= PVR_TIMER_STATE_ERROR;
	    p_RecJob->updateJobEntry(entry);
	    return NULL;
	}
	
	string strCommandLog = g_ffmpegPath+strParams;
	strCommand = g_ffmpegPath;
	XBMC->Log(LOG_NOTICE,"Starting ffmpeg: %s",strCommandLog.c_str());
    }
    
    //POSIX
    es.set_binary_mode(exec_stream_t::s_out);
    es.set_wait_timeout(exec_stream_t::s_in,g_streamTimeout*1000);
    es.set_wait_timeout(exec_stream_t::s_out,g_streamTimeout*1000);
    es.start( strCommand , strParams);

    XBMC->Log(LOG_NOTICE,"Set stream timeout: %d",g_streamTimeout);
    void *fileHandle;
    fileHandle = XBMC->OpenFileForWrite(videoFile.c_str(), true);
    
    string buffer;
    //int bytes_read;
    streamsize bytes_read;
    bool startTransmission = false;
    bool firstKilobyteReaded = false;
    time_t last_readed = time(NULL);
    while(true)
    {
	string buff;
	try {
	    getline( es.out(bytes_read), buff,'\n' ).good();
	    buffer = buffer+buff+"\n";
	    length=length+buff.size();
	    last_readed = time(NULL);
	    
	    if (startTransmission == false)
	    {
		t_startRecTime = time(NULL);
		startTransmission = true;
	    }
	    
	    if (firstKilobyteReaded==false)
	    {
		int loop;
		int pos = -1;
		int loop_end = buffer.size()-8-sizeof(double);
		for (loop=0; loop<loop_end;loop++)
		{
		    if (buffer[loop]=='d' && buffer[loop+1]=='u' && buffer[loop+2]=='r' && buffer[loop+3]=='a' && buffer[loop+4]=='t' && buffer[loop+5]=='i' && buffer[loop+6]=='o' && buffer[loop+7]=='n')
			pos = loop;
		} 
		if (pos>=0)
		{
		    pos = pos+9;
		    union
		    {
			unsigned char dc[8];
			double dd;
		    } d;
		    d.dd = t_duration;
		    long one = 1;
		    //is isBigEndian?
		    if(!(*((char *)(&one))))
		    {
			buffer[pos+0] = d.dc[0]; buffer[pos+1] = d.dc[1]; buffer[pos+2] = d.dc[2]; buffer[pos+3] = d.dc[3]; buffer[pos+4] = d.dc[4]; buffer[pos+5] = d.dc[5]; buffer[pos+6] = d.dc[6]; buffer[pos+7] = d.dc[7];
		    }
		    else
		    {
			buffer[pos+0] = d.dc[7]; buffer[pos+1] = d.dc[6]; buffer[pos+2] = d.dc[5]; buffer[pos+3] = d.dc[4]; buffer[pos+4] = d.dc[3]; buffer[pos+5] = d.dc[2]; buffer[pos+6] = d.dc[1]; buffer[pos+7] = d.dc[0];
		    }
		    XBMC->Log(LOG_NOTICE, "Duration corrected");
		}    
	    }
	    
	    if (buffer.size()>1024)
	    {
		firstKilobyteReaded=true;
	    }
	    
	    if (firstKilobyteReaded==true)
	    {
		XBMC->WriteFile(fileHandle, buffer.c_str(), buffer.size());
		buffer.clear();
	    }
	}catch( std::exception const & e ) {
	    //nothing to read
	}
	
	p_RecJob->getJobEntry(t_iClientIndex, entry);
	
	now = time(NULL);
	if (now-last_readed>=g_streamTimeout)
	{
	    //something wrong - data not growing
	    es.close();
	    es.kill();
            XBMC->CloseFile(fileHandle);
               
            XBMC->Log(LOG_NOTICE, "Recording failed %s", entry.Timer.strTitle);
                   
            //Correct duration time
	    if (length>0)
	    {
		double duration = last_readed-t_startRecTime;
		CorrectDurationFLVFile (videoFile,duration);
	    }
	    else
	    {
		XBMC->DeleteFile(videoFile.c_str());
	    }
            entry.Status = PVR_STREAM_STOPPED;
            entry.Timer.state= PVR_TIMER_STATE_ERROR;
            p_RecJob->updateJobEntry(entry);
	    s_triggerTimerUpdate = true;
            return NULL;
	    
	}
        if (entry.Timer.endTime<time(NULL) || entry.Status==PVR_STREAM_IS_STOPPING || entry.Status==PVR_STREAM_STOPPED)
        {
	    es.close();
	    es.kill();
            XBMC->CloseFile(fileHandle);
                            
            XBMC->Log(LOG_NOTICE, "Recording stopped %s", entry.Timer.strTitle);
                   
            //Correct duration time
	    if (length>0)
	    {
		double duration = last_readed-t_startRecTime;
		CorrectDurationFLVFile (videoFile,duration);
	    }
	    else
	    {
		XBMC->DeleteFile(videoFile.c_str());
	    }
            entry.Status = PVR_STREAM_STOPPED;
            entry.Timer.state= PVR_TIMER_STATE_COMPLETED;
            p_RecJob->updateJobEntry(entry);
	    s_triggerTimerUpdate = true;
            return NULL;
        }
    }
    es.close();
    es.kill();
    XBMC->CloseFile(fileHandle);
    time_t end_time = time(NULL);
    //Correct duration time
    if (length>0)
    {
	double duration = end_time-t_startRecTime;
	CorrectDurationFLVFile (videoFile,duration);
    }
    else
    {
	//File is empty
	XBMC->DeleteFile(videoFile.c_str());
    }
    entry.Status = PVR_STREAM_STOPPED;
    entry.Timer.state= PVR_TIMER_STATE_COMPLETED;
    p_RecJob->updateJobEntry(entry);
    s_triggerTimerUpdate = true;

    return NULL;
	
}


src/client.cpp
/*
 *      Copyright (C) 2013-2015 Anton Fedchin
 *      http://github.com/afedchin/xbmc-addon-iptvsimple/
 *
 *      Copyright (C) 2011 Pulse-Eight
 *      http://www.pulse-eight.com/
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with XBMC; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */


#include "client.h"
#include "kodi/xbmc_pvr_dll.h"
#include "kodi/libKODI_guilib.h"
#include "PVRIptvData.h"
#include "p8-platform/util/util.h"

using namespace ADDON;

#include "PVRRecorder.h"


#ifdef TARGET_WINDOWS
#define snprintf _snprintf
#endif

bool           m_bCreated       = false;
ADDON_STATUS   m_CurStatus      = ADDON_STATUS_UNKNOWN;
PVRIptvData   *m_data           = NULL;
PVRRecorder   *m_recorder   = NULL;
bool           m_bIsPlaying     = false;
PVRIptvChannel m_currentChannel;

/* User adjustable settings are saved here.
 * Default values are defined inside client.h
 * and exported to the other source files.
 */
std::string g_strUserPath   = "";
std::string g_strClientPath = "";

CHelper_libXBMC_addon *XBMC = NULL;
CHelper_libXBMC_pvr   *PVR  = NULL;

std::string g_strTvgPath    = "";
std::string g_strM3UPath    = "";
std::string g_strLogoPath   = "";
int         g_iEPGTimeShift = 0;
int         g_iStartNumber  = 1;
bool        g_bTSOverride   = true;
bool        g_bCacheM3U     = false;
bool        g_bCacheEPG     = false;
int         g_iEPGLogos     = 0;
std::string g_recordingsPath   = "";
std::string g_ffmpegPath   = "";
std::string g_ffmpegParams   = "";
std::string g_rtmpdumpPath   = "";
std::string g_fileExtension = "";
int         g_streamTimeout = 60;
int         g_streamQuality = 1;

extern std::string PathCombine(const std::string &strPath, const std::string &strFileName)
{
  std::string strResult = strPath;
  if (strResult.at(strResult.size() - 1) == '\\' ||
      strResult.at(strResult.size() - 1) == '/') 
  {
    strResult.append(strFileName);
  }
  else 
  {
    strResult.append("/");
    strResult.append(strFileName);
  }

  return strResult;
}

extern std::string GetClientFilePath(const std::string &strFileName)
{
  return PathCombine(g_strClientPath, strFileName);
}

extern std::string GetUserFilePath(const std::string &strFileName)
{
  return PathCombine(g_strUserPath, strFileName);
}

extern "C" {

void ADDON_ReadSettings(void)
{
  char buffer[1024];
  int iPathType = 0;
  if (!XBMC->GetSetting("m3uPathType", &iPathType)) 
  {
    iPathType = 1;
  }
  if (iPathType)
  {
    if (XBMC->GetSetting("m3uUrl", &buffer)) 
    {
      g_strM3UPath = buffer;
    }
    if (!XBMC->GetSetting("m3uCache", &g_bCacheM3U))
    {
      g_bCacheM3U = true;
    }
  }
  else
  {
    if (XBMC->GetSetting("m3uPath", &buffer)) 
    {
      g_strM3UPath = buffer;
    }
    g_bCacheM3U = false;
  }
  if (!XBMC->GetSetting("startNum", &g_iStartNumber)) 
  {
    g_iStartNumber = 1;
  }
  if (!XBMC->GetSetting("epgPathType", &iPathType)) 
  {
    iPathType = 1;
  }
  if (iPathType)
  {
    if (XBMC->GetSetting("epgUrl", &buffer)) 
    {
      g_strTvgPath = buffer;
    }
    if (!XBMC->GetSetting("epgCache", &g_bCacheEPG))
    {
      g_bCacheEPG = true;
    }
  }
  else
  {
    if (XBMC->GetSetting("epgPath", &buffer)) 
    {
      g_strTvgPath = buffer;
    }
    g_bCacheEPG = false;
  }
  float fShift;
  if (XBMC->GetSetting("epgTimeShift", &fShift))
  {
    g_iEPGTimeShift = (int)(fShift * 3600.0); // hours to seconds
  }
  if (!XBMC->GetSetting("epgTSOverride", &g_bTSOverride))
  {
    g_bTSOverride = true;
  }
  if (!XBMC->GetSetting("logoPathType", &iPathType)) 
  {
    iPathType = 1;
  }
  if (XBMC->GetSetting(iPathType ? "logoBaseUrl" : "logoPath", &buffer)) 
  {
    g_strLogoPath = buffer;
  }

  // Logos from EPG
  if (!XBMC->GetSetting("logoFromEpg", &g_iEPGLogos))
    g_iEPGLogos = 0;
  
  if (XBMC->GetSetting("recordingsPath", &buffer)) {
    g_recordingsPath = buffer;
  }
  
  if (XBMC->GetSetting("ffmpegPath", &buffer)) {
    g_ffmpegPath = buffer;
  }
  
  if (XBMC->GetSetting("ffmpegParams", &buffer)) {
    g_ffmpegParams = buffer;
  }
  
  if (XBMC->GetSetting("rtmpdumpPath", &buffer)) {
    g_rtmpdumpPath = buffer;
  }
  
  if (XBMC->GetSetting("fileExtension", &buffer)) {
    g_fileExtension = buffer;
  }
  
  int streamTimeout;
  if (XBMC->GetSetting("streamTimeout", &streamTimeout))
  {
    g_streamTimeout = streamTimeout;
  }
  
  int streamQuality;
  if (XBMC->GetSetting("streamQuality", &streamQuality))
  {
    g_streamQuality = streamQuality;
  }
}

ADDON_STATUS ADDON_Create(void* hdl, void* props)
{
  if (!hdl || !props)
  {
    return ADDON_STATUS_UNKNOWN;
  }

  PVR_PROPERTIES* pvrprops = (PVR_PROPERTIES*)props;

  XBMC = new CHelper_libXBMC_addon;
  if (!XBMC->RegisterMe(hdl))
  {
    SAFE_DELETE(XBMC);
    return ADDON_STATUS_PERMANENT_FAILURE;
  }

  PVR = new CHelper_libXBMC_pvr;
  if (!PVR->RegisterMe(hdl))
  {
    SAFE_DELETE(PVR);
    SAFE_DELETE(XBMC);
    return ADDON_STATUS_PERMANENT_FAILURE;
  }

  XBMC->Log(LOG_DEBUG, "%s - Creating the PVR IPTV Simple add-on", __FUNCTION__);

  m_CurStatus     = ADDON_STATUS_UNKNOWN;
  g_strUserPath   = pvrprops->strUserPath;
  g_strClientPath = pvrprops->strClientPath;

  if (!XBMC->DirectoryExists(g_strUserPath.c_str()))
  {
#ifdef TARGET_WINDOWS
    CreateDirectory(g_strUserPath.c_str(), NULL);
#else
    XBMC->CreateDirectory(g_strUserPath.c_str());
#endif
  }

  ADDON_ReadSettings();

  m_data = new PVRIptvData;
  m_recorder = new PVRRecorder();
  m_CurStatus = ADDON_STATUS_OK;
  m_bCreated = true;

  return m_CurStatus;
}

ADDON_STATUS ADDON_GetStatus()
{
  return m_CurStatus;
}

void ADDON_Destroy()
{
  delete m_data;
  m_bCreated = false;
  m_CurStatus = ADDON_STATUS_UNKNOWN;
}

bool ADDON_HasSettings()
{
  return true;
}

unsigned int ADDON_GetSettings(ADDON_StructSetting ***sSet)
{
  return 0;
}

ADDON_STATUS ADDON_SetSetting(const char *settingName, const void *settingValue)
{
  // reset cache and restart addon 

  std::string strFile = GetUserFilePath(M3U_FILE_NAME);
  if (XBMC->FileExists(strFile.c_str(), false))
  {
#ifdef TARGET_WINDOWS
    DeleteFile(strFile.c_str());
#else
    XBMC->DeleteFile(strFile.c_str());
#endif
  }

  strFile = GetUserFilePath(TVG_FILE_NAME);
  if (XBMC->FileExists(strFile.c_str(), false))
  {
#ifdef TARGET_WINDOWS
    DeleteFile(strFile.c_str());
#else
    XBMC->DeleteFile(strFile.c_str());
#endif
  }

  return ADDON_STATUS_NEED_RESTART;
}

void ADDON_Stop()
{
}

void ADDON_FreeSettings()
{
}

void ADDON_Announce(const char *flag, const char *sender, const char *message, const void *data)
{
}

/***********************************************************
 * PVR Client AddOn specific public library functions
 ***********************************************************/

const char* GetPVRAPIVersion(void)
{
  static const char *strApiVersion = XBMC_PVR_API_VERSION;
  return strApiVersion;
}

const char* GetMininumPVRAPIVersion(void)
{
  static const char *strMinApiVersion = XBMC_PVR_MIN_API_VERSION;
  return strMinApiVersion;
}

const char* GetGUIAPIVersion(void)
{
  return KODI_GUILIB_API_VERSION;
}

const char* GetMininumGUIAPIVersion(void)
{
  return KODI_GUILIB_MIN_API_VERSION;
}

PVR_ERROR GetAddonCapabilities(PVR_ADDON_CAPABILITIES* pCapabilities)
{
  pCapabilities->bSupportsEPG             = true;
  pCapabilities->bSupportsTV              = true;
  pCapabilities->bSupportsRadio           = true;
  pCapabilities->bSupportsChannelGroups   = true;
  pCapabilities->bSupportsRecordings      = true;
  pCapabilities->bSupportsTimers          = true;

  return PVR_ERROR_NO_ERROR;
}

const char *GetBackendName(void)
{
  static const char *strBackendName = "IPTV Simple PVR Add-on";
  return strBackendName;
}

const char *GetBackendVersion(void)
{
  static std::string strBackendVersion = XBMC_PVR_API_VERSION;
  return strBackendVersion.c_str();
}

const char *GetConnectionString(void)
{
  static std::string strConnectionString = "connected";
  return strConnectionString.c_str();
}

const char *GetBackendHostname(void)
{
  return "";
}

PVR_ERROR GetDriveSpace(long long *iTotal, long long *iUsed)
{
  *iTotal = 0;
  *iUsed  = 0;
  return PVR_ERROR_NO_ERROR;
}

PVR_ERROR GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd)
{
  if (m_data)
    return m_data->GetEPGForChannel(handle, channel, iStart, iEnd);

  return PVR_ERROR_SERVER_ERROR;
}

int GetChannelsAmount(void)
{
  if (m_data)
    return m_data->GetChannelsAmount();

  return -1;
}

PVR_ERROR GetChannels(ADDON_HANDLE handle, bool bRadio)
{
  if (m_data)
    return m_data->GetChannels(handle, bRadio);

  return PVR_ERROR_SERVER_ERROR;
}

bool OpenLiveStream(const PVR_CHANNEL &channel)
{
  if (m_data)
  {
    CloseLiveStream();

    if (m_data->GetChannel(channel, m_currentChannel))
    {
      m_bIsPlaying = true;
      return true;
    }
  }

  return false;
}

void CloseLiveStream(void)
{
  m_bIsPlaying = false;
}

int GetCurrentClientChannel(void)
{
  return m_currentChannel.iUniqueId;
}

bool SwitchChannel(const PVR_CHANNEL &channel)
{
  CloseLiveStream();

  return OpenLiveStream(channel);
}

PVR_ERROR GetStreamProperties(PVR_STREAM_PROPERTIES* pProperties)
{
  return PVR_ERROR_NOT_IMPLEMENTED;
}

int GetChannelGroupsAmount(void)
{
  if (m_data)
    return m_data->GetChannelGroupsAmount();

  return -1;
}

PVR_ERROR GetChannelGroups(ADDON_HANDLE handle, bool bRadio)
{
  if (m_data)
    return m_data->GetChannelGroups(handle, bRadio);

  return PVR_ERROR_SERVER_ERROR;
}

PVR_ERROR GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROUP &group)
{
  if (m_data)
    return m_data->GetChannelGroupMembers(handle, group);

  return PVR_ERROR_SERVER_ERROR;
}

PVR_ERROR SignalStatus(PVR_SIGNAL_STATUS &signalStatus)
{
  snprintf(signalStatus.strAdapterName, sizeof(signalStatus.strAdapterName), "IPTV Simple Adapter 1");
  snprintf(signalStatus.strAdapterStatus, sizeof(signalStatus.strAdapterStatus), "OK");

  return PVR_ERROR_NO_ERROR;
}

PVR_ERROR AddTimer(const PVR_TIMER &timer)
{
  return m_recorder->AddTimer (timer);
}

//PVR_ERROR DeleteTimer(const PVR_TIMER &timer, bool bForceDelete, bool bDeleteScheduled)
PVR_ERROR DeleteTimer(const PVR_TIMER& timer, bool bForceDelete)
{
  return m_recorder->DeleteTimer (timer,bForceDelete);
}

PVR_ERROR UpdateTimer(const PVR_TIMER &timer)
{
  return m_recorder->UpdateTimer (timer);
}

PVR_ERROR GetTimers(ADDON_HANDLE handle)
{
  return m_recorder->GetTimers(handle);
}

int GetTimersAmount(void)
{
  int job = m_recorder->GetTimersAmount();
  return job; 
}

/** UNUSED API FUNCTIONS */
const char * GetLiveStreamURL(const PVR_CHANNEL &channel)  { return ""; }
bool CanPauseStream(void) { return false; }
int GetRecordingsAmount(bool deleted) { return -1; }
PVR_ERROR GetRecordings(ADDON_HANDLE handle, bool deleted) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR OpenDialogChannelScan(void) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR CallMenuHook(const PVR_MENUHOOK &menuhook, const PVR_MENUHOOK_DATA &item) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR DeleteChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR RenameChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR MoveChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR OpenDialogChannelSettings(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR OpenDialogChannelAdd(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
bool OpenRecordedStream(const PVR_RECORDING &recording) { return false; }
void CloseRecordedStream(void) {}
int ReadRecordedStream(unsigned char *pBuffer, unsigned int iBufferSize) { return 0; }
long long SeekRecordedStream(long long iPosition, int iWhence /* = SEEK_SET */) { return 0; }
long long PositionRecordedStream(void) { return -1; }
long long LengthRecordedStream(void) { return 0; }
void DemuxReset(void) {}
void DemuxFlush(void) {}
int ReadLiveStream(unsigned char *pBuffer, unsigned int iBufferSize) { return 0; }
long long SeekLiveStream(long long iPosition, int iWhence /* = SEEK_SET */) { return -1; }
long long PositionLiveStream(void) { return -1; }
long long LengthLiveStream(void) { return -1; }
PVR_ERROR DeleteRecording(const PVR_RECORDING &recording) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR RenameRecording(const PVR_RECORDING &recording) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR SetRecordingPlayCount(const PVR_RECORDING &recording, int count) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR SetRecordingLastPlayedPosition(const PVR_RECORDING &recording, int lastplayedposition) { return PVR_ERROR_NOT_IMPLEMENTED; }
int GetRecordingLastPlayedPosition(const PVR_RECORDING &recording) { return -1; }
PVR_ERROR GetRecordingEdl(const PVR_RECORDING&, PVR_EDL_ENTRY[], int*) { return PVR_ERROR_NOT_IMPLEMENTED; };
PVR_ERROR GetTimerTypes(PVR_TIMER_TYPE types[], int *size) { return PVR_ERROR_NOT_IMPLEMENTED; }
void DemuxAbort(void) {}
DemuxPacket* DemuxRead(void) { return NULL; }
unsigned int GetChannelSwitchDelay(void) { return 0; }
bool IsTimeshifting(void) { return false; }
void PauseStream(bool bPaused) {}
bool CanSeekStream(void) { return false; }
//bool SeekTime(int,bool,double*) { return false; }
bool SeekTime(double, bool, double*) { return false; }
void SetSpeed(int) {};
time_t GetPlayingTime() { return 0; }
time_t GetBufferTimeStart() { return 0; }
time_t GetBufferTimeEnd() { return 0; }
PVR_ERROR UndeleteRecording(const PVR_RECORDING& recording) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR DeleteAllRecordingsFromTrash() { return PVR_ERROR_NOT_IMPLEMENTED; }
}
